llvm/lib/CodeGen/MachinePipeliner.cpp

```diff
@@ -1326,48 +1326,7 @@ class HighRegisterPressureDetector {
   // Calculate the upper limit of each pressure set
   void computePressureSetLimit(const RegisterClassInfo &RCI) {
     for (unsigned PSet = 0; PSet < PSetNum; PSet++)
-      PressureSetLimit[PSet] = TRI->getRegPressureSetLimit(MF, PSet);
-
-    // We assume fixed registers, such as stack pointer, are already in use.
-    // Therefore subtracting the weight of the fixed registers from the limit of
-    // each pressure set in advance.
-    SmallDenseSet<Register, 8> FixedRegs;
```

Comment:
Not sure what a fixed register is, but it seems to not be the same as reserved. Seems like another hook with bad defaults

Reply:
Yeah, it is a bad API. Fixed registers are not reserved register... Fixed register are registers defined in `FixedRegisters` RegisterCategory by default...
The logic here seems to want to remove reserved registers @kasuga-fj.

Reply:
As I commented in https://github.com/llvm/llvm-project/pull/118787#discussion_r1877336858, I dared to replace it from `RegisterClassInfo::getRegPressureSetLimit` before (in #87312). This is because there is a duplication between the reserved registers handled inside `RegisterClassInfo::getRegPressureSetLimit` and the `FixedRegs` calculated here, so the `Limit >= Weight` assertion fails. If I remember correctly, in my case (AArch64), the `$ffr` triggers it. What I'd like to do here is to ignore registers that have "specific rolls" (e.g., stack pointer) and those where `TargetRegisterInfo::isFixedRegister` returns true looks appropriate in this case to me. Maybe I'm misunderstanding reserved and fixed registers, and there may be a better way to achieve what I want, but I'm not sure this change will work well.

Reply:
IMO isFixedRegister should be deleted. The intent seems to be isReserved || isConstantPhysREg

Reply:
Thanks for explanation! At least, after removing this part of code, the tests still pass? @kasuga-fj

Reply:
I ran tests and none of them failed. But that just means there are no correctness issues. This code is part of a heuristic decision, so performance degradation could happen in some cases (and I think the heuristics will be less accurate with this change).

Reply:
> I think the heuristics will be less accurate with this change

I don't think so. These "fixed" register are also "reserved" registers, and "reserved" registers contain runtime fixed registers like those are reserved via `-ffixed-xx` options.

Reply:
>  These "fixed" registers are also "reserved" registers,

Previously (where `RegisterClassInfo::getRegPressureSetLimit` was used here), #83437 caused the `Limit >= Weight` assertion to fail. This patch marks the `$ffr` register as reserved. I just tried to comment out the code that marks `$ffr` as reserved and run MachinePipeliner, and found that `FixedRegs` contains `$ffr`. It looks to me like there are some gaps between "reserved" and "fixed". Or maybe I misunderstood something (I don't know much about the LLVM API for registers).